crates/orderbook/src/ipfs.rs

```diff
@@ -0,0 +1,162 @@
+    /// be confusing:
+    ///
+    /// - The public ipfs.io gateway responds "504 Gateway Timeout" after 2
+    ///   minutes.
+    /// - The public cloudflare gateway responds "524" after 20 seconds.
+    /// - A private Pinata gateway responds "404 Not Found" after 2 minutes.
+    ///
+    /// This function treats all status codes except "200 OK" as errors and you
+    /// likely want to use it with a timeout.
+    pub async fn fetch(&self, cid: &str) -> Result<Vec<u8>> {
+        let url = self.prepare_url(cid);
+        let response = self.client.get(url).send().await.context("send")?;
+        let status = response.status();
+        let body = response.bytes().await.context("body")?;
+        match status {
+            StatusCode::OK => Ok(body.into()),
+            _ => {
+                let body_text = String::from_utf8_lossy(&body);
+                let body_text: &str = &body_text;
+                Err(anyhow!("status {status}, body {body_text:?}"))
+            }
+        }
+    }
+
+    fn prepare_url(&self, cid: &str) -> Url {
+        let mut url = self.base.clone();
+        let mut path = url.path_segments_mut().unwrap();
+        path.push("ipfs");
+        path.push(cid);
+        std::mem::drop(path);
+        if let Some(query) = &self.query {
+            url.set_query(Some(query.as_str()));
+        }
+        url
+    }
+}
+
+/// Tries to find full app data corresponding to the contract app data on IPFS.
+///
+/// A return value of `Some` indicates that either the old or new CID format was
+/// found on IPFS and points to valid utf-8.
+///
+/// A return value of `None` indicates that neither CID was found. This might be
+/// a temporary condition as IPFS is a decentralized network.
+pub async fn full_app_data_from_ipfs(
+    ipfs: &Ipfs,
+    contract_app_data: &AppDataHash,
+) -> Option<String> {
+    let old = old_app_data_cid(contract_app_data);
+    let new = new_app_data_cid(contract_app_data);
+    let fetch = |cid: String| async move {
+        let data = match ipfs.fetch(&cid).await {
+            Ok(data) => {
+                tracing::debug!("found full app data for {contract_app_data:?} at CID {cid}");
+                data
+            }
+            Err(err) => {
+                tracing::debug!("no full app data for {contract_app_data:?} at CID {cid}: {err:?}");
+                return None;
+            }
+        };
+        match String::from_utf8(data) {
+            Ok(data) => Some(data),
+            Err(_) => {
+                tracing::debug!("CID {cid} doesn't point to utf-8");
+                None
+            }
+        }
+    };
+    match futures::future::select(std::pin::pin!(fetch(old)), std::pin::pin!(fetch(new))).await {
+        Either::Left((complete, pending)) | Either::Right((complete, pending)) => {
+            if complete.is_some() {
+                complete
+            } else {
+                pending.await
+            }
+        }
+    }
```

Comment:
Looks like this could be replaced with [`select_ok`](https://docs.rs/futures/latest/futures/future/fn.select_ok.html). (with slight modifications)

Reply:
done